csharp/src/Google.Protobuf/CodedInputReader.cs

```diff
@@ -0,0 +1,879 @@
+            int maxCharLength = Encoding.UTF8.GetMaxCharCount(byteLength);
+            char[] charArray = ArrayPool<char>.Shared.Rent(maxCharLength);
+
+            try
+            {
+                int remainingByteLength = byteLength;
+                int initializedChars = 0;
+                while (remainingByteLength > 0)
+                {
+                    int bytesRead = Math.Min(remainingByteLength, reader.UnreadSpan.Length);
+                    remainingByteLength -= bytesRead;
+                    bool flush = remainingByteLength == 0;
+
+                    unsafe
+                    {
+                        fixed (byte* pUnreadSpan = &MemoryMarshal.GetReference(reader.UnreadSpan))
+                        fixed (char* pCharArray = &charArray[initializedChars])
+                        {
+                            initializedChars += decoder.GetChars(pUnreadSpan, bytesRead, pCharArray, charArray.Length - initializedChars, flush);
+                        }
+
+                        reader.Advance(bytesRead);
+                    }
+                }
+
+                string value = new string(charArray, 0, initializedChars);
+                return value;
+            }
+            finally
+            {
+                ArrayPool<char>.Shared.Return(charArray);
+            }
+        }
+
+        /// <summary>
+        /// Reads an embedded message field value from the input.
+        /// </summary>   
+        public void ReadMessage(IBufferMessage builder)
+        {
+            int length = ReadLength();
+            if (recursionDepth >= recursionLimit)
+            {
+                throw InvalidProtocolBufferException.RecursionLimitExceeded();
+            }
+            long oldLimit = PushLimit(length);
+            ++recursionDepth;
+            builder.MergeFrom(ref this);
+            CheckReadEndOfInputTag();
+            // Check that we've read exactly as much data as expected.
+            if (!ReachedLimit)
+            {
+                throw InvalidProtocolBufferException.TruncatedMessage();
+            }
+            --recursionDepth;
+            PopLimit(oldLimit);
+        }
+
+        /// <summary>
+        /// Reads an embedded group field from the input.
+        /// </summary>
+        public void ReadGroup(IBufferMessage builder)
+        {
+            if (recursionDepth >= recursionLimit)
+            {
+                throw InvalidProtocolBufferException.RecursionLimitExceeded();
+            }
+            ++recursionDepth;
+            builder.MergeFrom(ref this);
+            --recursionDepth;
+        }
+
+        /// <summary>
+        /// Reads a bytes field value from the input.
+        /// </summary>   
+        public ByteString ReadBytes()
+        {
+            int length = ReadLength();
+
+            if (length == 0)
+            {
+                return ByteString.Empty;
+            }
+
+            ThrowEndOfInputIfFalse(reader.Remaining >= length);
```

Comment:
similar to SkipLastField, you need to consult  the currentLimit too? Looks like without it can go over the allowed limit for current field without noticing?